python/cudf/cudf/core/index.py

```diff
@@ -434,6 +438,130 @@ def _concat(cls, objs):
+                # short-circuit and return a copy
+                to_concat = [self]
+
+            other = as_index(other)
+
+            if len(self) == 0:
+                to_concat = [other]
+
+            if len(self) and len(other):
+                if (this.dtype == "object" and other.dtype != "object") or (
+                    other.dtype == "object" and this.dtype != "object"
+                ):
+                    raise TypeError(
+                        "cudf does not support mixed types, please type-cast "
+                        "both index to same dtypes."
+                    )
+
+                if isinstance(self._values, cudf.core.column.NumericalColumn):
+                    if self.dtype != other.dtype:
+                        this, other = numeric_normalize_types(self, other)
+                to_concat = [this, other]
+
+        for obj in to_concat:
+            if not isinstance(obj, Index):
+                raise TypeError("all inputs must be Index")
+
+        return self._concat(to_concat)
+
+    def difference(self, other, sort=None):
+        """
+        Return a new Index with elements from the index that are not in
+        `other`.
+
+        This is the set difference of two Index objects.
+
+        Parameters
+        ----------
+        other : Index or array-like
+        sort : False or None, default None
+            Whether to sort the resulting index. By default, the
+            values are attempted to be sorted, but any TypeError from
+            incomparable elements is caught by cudf.
+            * None : Attempt to sort the result, but catch any TypeErrors
+              from comparing incomparable elements.
+            * False : Do not sort the result.
+
+        Returns
+        -------
+        difference : Index
+
+        Examples
+        --------
+        >>> import cudf
+        >>> idx1 = cudf.Index([2, 1, 3, 4])
+        >>> idx1
+        Int64Index([2, 1, 3, 4], dtype='int64')
+        >>> idx2 = cudf.Index([3, 4, 5, 6])
+        >>> idx2
+        Int64Index([3, 4, 5, 6], dtype='int64')
+        >>> idx1.difference(idx2)
+        Int64Index([1, 2], dtype='int64')
+        >>> idx1.difference(idx2, sort=False)
+        Int64Index([2, 1], dtype='int64')
+        """
+
+        other = as_index(other)
+
+        if self.dtype != other.dtype and (
+            cudf.utils.dtypes.is_string_dtype(self.dtype)
+            or cudf.utils.dtypes.is_string_dtype(other.dtype)
+        ):
+            if self.equals(other):
+                if cudf.utils.dtypes.is_string_dtype(self.dtype):
+                    return StringIndex([], dtype=self.dtype)
+                else:
+                    return self[:0]
+            else:
+                difference = self
+        else:
+            difference = self.join(other, how="leftanti")
+            if self.dtype != other.dtype:
+                difference = difference.astype(self.dtype)
+
+        if sort is None:
```

Comment:
Need to validate that `sort is in {None, False}` otherwise passing `sort=True` won't throw and won't sort.